lib/src/command.dart

```diff
@@ -98,37 +107,42 @@ abstract class PubCommand extends Command<int> {
 
   /// The first command in the command chain.
   Command get _topCommand {
-    var command = this;
-    while (command.parent != null) {
-      command = command.parent;
+    Command current = this;
+    while (true) {
+      var parent = current.parent;
+      if (parent == null) return current;
+      current = parent;
     }
-    return command;
   }
 
-  PubEmbeddableCommand get _pubEmbeddableCommand {
-    var command = this;
-    while (command != null && command is! PubEmbeddableCommand) {
+  PubEmbeddableCommand? get _pubEmbeddableCommand {
+    Command? command = this;
+    while (command is! PubEmbeddableCommand) {
+      if (command == null) {
+        return null;
+      }
       command = command.parent;
     }
+
     return command;
   }
 
-  PubTopLevel get _pubTopLevel {
-    return _pubEmbeddableCommand ?? (runner as PubCommandRunner);
+  PubTopLevel? get _pubTopLevel {
```

Comment:
Do you think we can make this be non-nullable?
I think we should never be without a top-level

Reply:
Can you see a way to get there?

Reply:
I think I would just do:
```
PubTopLevel get _pubTopLevel => _pubEmbeddableCommand ?? runner!;
```

Reply:
`PubTopLevel get _pubTopLevel => _pubEmbeddableCommand ?? runner as PubCommandRunner;`